### PR TITLE
chore(reth-network-p2p): Replace HeadersClient + BodiesClient with BlockClient

### PR DESCRIPTION
**Pull request by nkysg**
Partly towards https://github.com/paradigmxyz/reth/issues/10053

### CODE REVIEW

File: crates/net/p2p/src/lib.rs [hunk 1]
```diff
@@ -52,6 +52,6 @@ pub use headers::client::HeadersClient;
 pub use reputation::{Reputation, ReputationChange, ReputationChangeKind, ReputationChangeWeights};
 
 /// Helper trait that unifies network behaviour needed for fetching blocks.
-pub trait BlockClient: HeadersClient + BodiesClient + Unpin + Clone {}
```
Comment (emhane) [inline crates/net/p2p/src/lib.rs:55]: we don't want this. please revert.

**Review by emhane — CHANGES_REQUESTED** [review]

File: crates/net/p2p/src/lib.rs (hunk 1 above)
Comment (mattsse) [reply crates/net/p2p/src/lib.rs:55]: ah because we don't need really to enforce this in too many locations?

File: crates/net/p2p/src/lib.rs [hunk 2]
```diff
@@ -52,6 +52,6 @@ pub use headers::client::HeadersClient;
...
+pub trait BlockClient: HeadersClient + BodiesClient {}
 
-impl<T> BlockClient for T where T: HeadersClient + BodiesClient + Unpin + Clone {}
```
Comment (mattsse) [inline crates/net/p2p/src/lib.rs:57]: this is fine though

**Review by mattsse — COMMENTED** [review]
very nice

File: crates/net/p2p/src/lib.rs (hunk 1 above)
Comment (emhane) [reply crates/net/p2p/src/lib.rs:55]: we have a style already through out the code base of including autotraits as super traits of our traits

Comment (emhane) [reply crates/net/p2p/src/lib.rs:55]: regardless how often they are actually needed

File: crates/net/p2p/src/full_block.rs [hunk 3]
```diff
@@ -41,7 +42,7 @@ impl<Client> FullBlockClient<Client> {
 
 impl<Client> FullBlockClient<Client>
 where
-    Client: BodiesClient + HeadersClient + Clone,
+    Client: BlockClient + Clone,
```
Comment (SkandaBhat) [inline crates/net/p2p/src/full_block.rs:45]:
```suggestion
    Client: BlockClient,
```

Similar to this comment: https://github.com/paradigmxyz/reth/pull/10059#discussion_r1702863335

**Review by emhane — APPROVED** [review]

**Comment by emhane** [comment]
not sure why the test fails, will open an issue for flaky test

**Pull request merged**